resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -742,29 +855,39 @@ private[spark] class Client(
     val servicePort = service.getSpec.getPorts.asScala
       .filter(_.getName == SUBMISSION_SERVER_PORT_NAME)
       .head.getNodePort
-    val nodeUrls = kubernetesClient.nodes.list.getItems.asScala
-      .filterNot(node => node.getSpec.getUnschedulable != null &&
-        node.getSpec.getUnschedulable)
-      .flatMap(_.getStatus.getAddresses.asScala)
-      // The list contains hostnames, internal and external IP addresses.
-      // (https://kubernetes.io/docs/admin/node/#addresses)
-      // we want only external IP addresses and legacyHostIP addresses in our list
-      // legacyHostIPs are deprecated and will be removed in the future.
-      // (https://github.com/kubernetes/kubernetes/issues/9267)
-      .filter(address => address.getType == "ExternalIP" || address.getType == "LegacyHostIP")
-      .map(address => {
-        s"$urlScheme://${address.getAddress}:$servicePort"
-      }).toSet
+    val serverBasePath = s"$kubernetesAppId/$SUBMISSION_SERVER_PATH_COMPONENT"
+    val nodeUrls = ingressBasePath.map(ingressBase => {
+      Set(s"$urlScheme://$ingressBase/$serverBasePath")
+    }).getOrElse {
+      kubernetesClient.nodes.list.getItems.asScala
+        .filterNot(node => node.getSpec.getUnschedulable != null &&
+          node.getSpec.getUnschedulable)
+        .flatMap(_.getStatus.getAddresses.asScala)
+        // The list contains hostnames, internal and external IP addresses.
+        // (https://kubernetes.io/docs/admin/node/#addresses)
+        // we want only external IP addresses and legacyHostIP addresses in our list
+        // legacyHostIPs are deprecated and will be removed in the future.
+        // (https://github.com/kubernetes/kubernetes/issues/9267)
+        .filter(address => address.getType == "ExternalIP" || address.getType == "LegacyHostIP")
+        .map(address => {
+          s"$urlScheme://${address.getAddress}:$servicePort/$serverBasePath"
+        }).toSet
+    }
     require(nodeUrls.nonEmpty, "No nodes found to contact the driver!")
     val (trustManager, sslContext): (X509TrustManager, SSLContext) =
       if (driverSubmitSslOptions.enabled) {
         buildSslConnectionConfiguration(driverSubmitSslOptions)
       } else {
         (null, SSLContext.getDefault)
       }
+    val maxRetriesPerServer = if (useIngress) {
+      SUBMISSION_CLIENT_RETRIES_INGRESS
```

Comment:
@foxish is there perhaps a better way to do this? It turns out that the Ingress resource takes longer to become ready than the NodePort endpoints. Furthermore, the `Watch` on the Ingress doesn't necessarily catch whether or not the Ingress has added the endpoint.